overlord/snapstate/handlers.go

```diff
@@ -310,20 +323,120 @@ func (m *SnapManager) installOneBaseOrRequired(st *state.State, snapName string,
 	return ts, err
 }
 
-func (m *SnapManager) installPrereqs(t *state.Task, base string, prereq []string, userID int, tm timings.Measurer) error {
+// updates a prerequisite, if it's not providing a content interface that a plug expects it to
+func updatePrereqIfOutdated(t *state.Task, snapName string, contentAttrs []string, userID int, flags Flags) (*state.TaskSet, error) {
+	if len(contentAttrs) == 0 {
+		return nil, nil
+	}
+
+	st := t.State()
+
+	// check if the default provider has all expected content tags
+	if ok, err := hasAllContentAttrs(st, snapName, contentAttrs); err != nil {
+		return nil, err
+	} else if ok {
+		return nil, nil
+	}
+
+	if ok, err := shouldSkipToAvoidConflict(t, snapName); err != nil {
+		return nil, err
+	} else if ok {
+		return nil, nil
+	}
+
+	// default provider is missing some content tags (likely outdated) so update it
+	ts, err := Update(st, snapName, nil, userID, flags)
+	if err != nil {
+		if conflErr, ok := err.(*ChangeConflictError); ok {
+			// there's already an update for the same snap in this change,
+			// just skip this one
+			if conflErr.ChangeID == t.Change().ID() {
```

Comment:
AFAIU (but I may be missing something) we generally avoid conflict with shouldSkipToAvoidConflict() above, so I'm not quite clear what are the exact circumstances we can hit it here (even considering that Update unlocks the state around store call) - but I can see we definitely hit this case in the unit tests. I think both this and the call to shouldSkipToAvoidConflict above should have a comment explaining the conflicts, and why can we hit it again here.

Reply:
It's been a while since the previous check was discussed so I may be forgetting something but, looking at it, it seems like what `shouldSkipToAvoidConflict` does is a subset of the conflict checking in Update so the first could be removed. Maybe @pedronis can correct me if I'm wrong

Reply:
Update talks to the store first, so I think this is fine as is